File: lesson_12/structs_ts/src/stack.ts
import { ListNode } from './list_node.js';

export class Stack {
  private top: ListNode | undefined;

  constructor() {
    this.top = undefined;
  }

  push(value: number): void {
    const newNode = new ListNode(value);
    newNode.next = this.top;
    this.top = newNode;
  }

  pop(): number | undefined {
    let topmostValue = undefined;
    if (this.isEmpty())
       throw new Error('No value to pop');
    else 
      topmostValue = this.top?.val;  
      this.top = this.top?.next;
      return topmostValue
  }

  peek(): number | null {
      if (this.top == undefined) 
        throw new Error('No value to pop');
      else
       return this.top.val;
  }

  isEmpty(): boolean {
    return (this.top == null);
  }  
}

File: lesson_12/structs_java/structs_app/src/main/java/com/codedifferently/lesson12/Stack.java
package com.codedifferently.lesson12;

public class Stack {
  private ListNode top;

  public Stack() {
    this.top = null;
  }

  public void push(int value) {
    ListNode newNode = new ListNode(value);
    newNode.next = top;
    top = newNode;
  }

  public int pop() {
    int topmostValue = 0;
    if (isEmpty()) {
      throw new IllegalStateException("Stack is empty");
    } else {
      topmostValue = top.val;
      top = top.next;
      return topmostValue;
    }
  }

  public int peek() {
    if (isEmpty()) {
      throw new IllegalStateException("Stack is empty, no value to peek");
    }
    return top.val;
  }

  public boolean isEmpty() {
    return top == null;
  }
}
